Issue #1031: Enable editing of DS attributes on allocation

katelynienaber opened the pull request:
## Proposed changes

This is an enhancement, linked to [issue #1031](https://github.com/zowe/vscode-extension-for-zowe/issues/1031).

It allows you to edit a data set's attributes before allocation, following the scheme presented in that issue and discussed with @jelaplan and @VitGottwald:

![changeParamsDS](https://user-images.githubusercontent.com/58596139/100616359-4cc35200-3319-11eb-9692-c5e07596d035.gif)

## Release Notes

Milestone: 2.0 (?)
Changelog: Enabled editing of DS attributes on allocation

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `npm run vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

zdmullen commented:
I've just gone through this once and one major thing I noticed is that  "Directory Blocks" is in the list of attributes twice, and this pushes the "+Allocate Data Set" down so that I have to scroll to see it. What that issue got me thinking about is whether a user may just want to edit one attribute and then be able to "+ Allocate Data Set". Was this part of the original discussion, that we should allow them to allocate after editing just one attribute so they don't have to go back to the full list and select the "+ Allocate..." part?

katelynienaber commented:
> I've just gone through this once and one major thing I noticed is that "Directory Blocks" is in the list of attributes twice, and this pushes the "+Allocate Data Set" down so that I have to scroll to see it. What that issue got me thinking about is whether a user may just want to edit one attribute and then be able to "+ Allocate Data Set". Was this part of the original discussion, that we should allow them to allocate after editing just one attribute so they don't have to go back to the full list and select the "+ Allocate..." part?

So doing it this way would give 2 possible flows (I guess):
- User selects an item to edit...user edits...user hits enter to allocate
- User selects an item to edit...user edits...user arrows down to "Back" option, then hits enter to go back to full list

This would make allocation _more_ inconvenient for people editing many attributes. Because if you want to edit more attributes, you will have to arrow down and go "Back" for every option you edit.

Idk maybe John or @jellypuno or @VitGottwald can put in here...does anyone know if people most frequently edit one attribute or more than one?

_Also I fixed that double item in the list @zdmullen, and changed the item order in the big list. Now if a user enters an attribute value and hits Enter, they can hit Enter one more time and it will allocate automatically._

JillieBeanSim reviewed (changes requested):
@katelynienaber this is really cool, I do have a few questions about what is presented to the users in the parameters list:
1. During create should we offer the option to change the node label since the data set name was just entered before hand?
2. If we do decide to keep the option to rename during this, would we want to use more familiar labels like `Data Set Name` instead of `Node Type`?
3. Another thing with labels, `Node Type` and `DSN type` are presented in the list where `Node Type` is holding the value, but shouldn't `DSN Type` hold the value and possibly have the long name `Data Set Type`,  and `Node Type` not be in the list as it isn't a traditional label when allocating a data set

Functionally this works well, just requesting a few polish items to follow traditional labels for the attributes.

One other thing I notice when running integration tests is that the tests seem to hang on the Creating Data Sets and Members tests (tried multiple times with same result).  I wonder if this hang is caused due to changes in the UI for this new functionality.  I know these tests are inconsistent across users that run them, so if someone else could run them and not get the hang I experienced it could just be me. 
![image](https://user-images.githubusercontent.com/49491949/104345592-88cb0480-54cc-11eb-92e3-dbb42b952e4f.png)

katelynienaber commented:
> 1. During create should we offer the option to change the node label since the data set name was just entered before hand?

I would still leave it, in case people make a typo or something

> 2. If we do decide to keep the option to rename during this, would we want to use more familiar labels like `Data Set Name` instead of `Node Label`?

I did change this :) It makes sense.

> 3. Another thing with labels, `Node Type` and `DSN type` are presented in the list where `Node Type` is holding the value, but shouldn't `DSN Type` hold the value and possibly have the long name `Data Set Type`,  and `Node Type` not be in the list as it isn't a traditional label when allocating a data set

I am getting these attributes from the object `zowe.ICreateDataSetOptions`, so I would like to include both. But I made some changes to the names to add clarity, let me know if this is adequate:
![image](https://user-images.githubusercontent.com/58596139/104712490-e5bff800-5722-11eb-816b-e80cd66e342a.png)

> One other thing I notice when running integration tests is that the tests seem to hang on the Creating Data Sets and Members tests (tried multiple times with same result).

This was an issue related to the new code...they are running okay for me now, please confirm!

jellypuno commented on packages/zowe-explorer/__tests__/__integration__/extension.integration.test.ts:
```diff
@@ -152,18 +156,28 @@ describe("Extension Integration Tests", () => {
 
+            // 2nd step: User selects DS type
+            const quickPickStub = sandbox.stub(vscode.window, "showQuickPick");
+            quickPickStub.onCall(0).returns("Data Set Sequential");
```
If possible can you change this to a foreach statement? The last PR that I opened has magic numbers error. I can see that yours doesn't have one but it would be great if we could prevent it in the future. I did it like this:

```
const returnVals = ["testUser", "testPass", "basePath", "encoding", "timeOut"];
        returnVals.forEach((value, index) => {
            showInputStub.onCall(index).returns(value);
        });
```

jellypuno commented on packages/zowe-explorer/src/dataset/actions.ts:
```diff
@@ -297,31 +298,38 @@ export async function openPS(
             break;
         case localize("createFile.dataSetSequential", "Data Set Sequential"):
             typeEnum = zowe.CreateDataSetTypeEnum.DATA_SET_SEQUENTIAL;
+            cliDefaultsKey = "SEQUENTIAL";
```
Do you think it is good to put this in globals? Maybe we can re-use it in future enhancements?

katelynienaber replied:
I am not sure if this is useful...the `cliDefaultsKey` is just the `CreateDataSetTypeEnum`, but without `DATA_SET_`

ie. if the `CreateDataSetTypeEnum` is `DATA_SET_BINARY`, then the `cliDefaultsKey` is `BINARY`...

So it feels a little pointless to me to make a whole new enum just to remove the `DATA_SET_` part of `CreateDataSetTypeEnum`

The CLI is irritating me a bit with all their different names for the same things. `CreateDataSetTypeEnum` should just match the keys used in the CLI to identify the default settings (keys used in `zowe.CreateDefaults` file). We are having the same issue with sending profile values to Imperative as CLI commands. I think they need to standardize their naming procedures, and we shouldn't be making a bunch of different data structures that only have slight differences in their values to accommodate all these slightly different names across CLI and Imperative.

Rant over. Maybe I will bring this up in standup.

jellypuno replied:
ok. I understand. Do you think trim would be useful since you know that you want to remove DATA_SET_? like this:

const session = node.label.trim().substring(node.label.trim().indexOf("[") + 1, node.label.trim().indexOf("]"));

jellypuno replied:
There's an example in C:\Users\jp669971\Documents\GitHub\Community\vscode-extension-for-zowe\packages\zowe-explorer\src\dataset\actions.ts. Line 989

katelynienaber replied:
The problem is that I can't get the key from the enum. Ex. I don't have DATA_SET_SEQUENTIAL as a string, I have just the number: 4.

If you look here, since it is a `const enum`, the key is removed completely on compilation...so it doesn't even exist there for me to get ahold of: [link to docs](https://www.typescriptlang.org/docs/handbook/enums.html#const-enums)

Maybe it makes sense to add it as an enum for this reason, but I still think it's a bit overkill P:

katelynienaber replied:
I could add an enum like this:
```
CreateDataSetTypeKeysEnum {
    0 = DATA_SET_BINARY,
    1 = DATA_SET_C,
    2 = DATA_SET_CLASSIC,
    3 = DATA_SET_PARTITIONED,
    4 = DATA_SET_SEQUENTIAL
}

```
And that would at least be more useful, in that we can take the `CreateDataSetTypeEnum` and reverse it to get the keys when we need them...but it is sort of a hack to get around their `const enum` definition, and it makes me wonder: why did they define it as `const enum` in the first place? Maybe they can remove it? Maybe we shouldn't go around it?

jellypuno replied:
I think that "hack" is the approach that I'm gonna go for. But only if you find it very useful. 

> why did they define it as const enum in the first place? 
That's a good question. We can open an enhancement issue for this. I pretty sure they didn't expect anybody to get the description of the dataset. So this might be the first time.

jellypuno commented on packages/zowe-explorer/src/globals.ts:
```diff
@@ -54,6 +54,116 @@ export const ICON_STATE_OPEN = "open";
 export const ICON_STATE_CLOSED = "closed";
 export const VSCODE_APPNAME: string[] = ["Visual Studio Code", "VSCodium"];
 export const ROOTPATH = path.join(__dirname, "..", "..");
+export const DATA_SET_PROPERTIES = [
```
Can you import the descriptions here: C:\Users\jp669971\Documents\GitHub\Community\vscode-extension-for-zowe\node_modules\@zowe\cli\lib\zosfiles\-strings-\en.d.ts

jellypuno replied:
And also, I think it needs to be localize for webpack

katelynienaber replied:
How do I do this?

jellypuno replied:
Maybe we should talk to CLI team about it but they did it in this way : https://github.com/zowe/zowe-cli/blob/a9b40537e432935fda7b03e27467bb854cf3db2e/packages/cli/src/zosfiles/hMigrate/HMigrate.definition.ts#L18

jellypuno replied:
Or maybe @lauren-li and @JillieBeanSim can give us some hints on how to localize these properties

katelynienaber replied:
I have done the localization but I am not sure how to import them into that file...so yeah it would be cool to ask someone about how to use that file :)

jellypuno replied:
The first part is right, I would like to import that file and use it. And I'm wondering if we could localize it without re-writing it in our codebase or can we use their own localization? (Maybe not 😢 )

lauren-li replied:
@jellypuno Assuming you mean that we should import descriptions from `@zowe\cli\lib\zosfiles-strings-\en.js` and display them localized in Zowe Explorer (as @JillieBeanSim described), I took a look at [this file](https://github.com/zowe/zowe-cli/blob/a9b40537e432935fda7b03e27467bb854cf3db2e/packages/cli/src/zosfiles/-strings-/en.ts ) in the Zowe CLI repo, and it seems we will need to do some JSON processing, especially since the keys there don't match the ones we need.

If we process the JSON and store their values in variables in Zowe Explorer, then I believe we should just be able to localize the values of those variables per our usual localization method.

JillieBeanSim replied:
so we would need to create a schema that could be used to process the JSON and could set a way to localize all Description fields?

lauren-li replied:
That might be one way to do it, although perhaps in a separate PR.

jellypuno replied:
I agree. Thank you for your guidance Lauren and Billie. I appreciate it 😊

@katelynienaber Just use the file for now and then the localization will be handled on a different PR. 😉

lauren-li replied:
I just had the thought that if we are using the imported strings file from `@zowe`, we'll be able to pull in the latest updated strings, but that means we probably won't be able to localize the strings ourselves. 🤔 But perhaps we could set some variable to update the path to the import based on the user's VS Code language setting, and use whatever localization is provided by `@zowe` (although currently, this only appears to be English).

jellypuno commented:
For the newly created datasets, make sure it is deleted in the posttest:integration script

katelynienaber commented:
> For the newly created datasets, make sure it is deleted in the posttest:integration script

Done

jellypuno commented:
This is not an issue in this PR but make sure that the dataset name is not too long. It has a limit of 44 chars only. For more reference: https://www.ibm.com/support/knowledgecenter/zosbasics/com.ibm.zos.zjcl/zjclt_dstypesNnames.htm

katelynienaber commented:
> This is not an issue in this PR but make sure that the dataset name is not too long. It has a limit of 44 chars only. For more reference: https://www.ibm.com/support/knowledgecenter/zosbasics/com.ibm.zos.zjcl/zjclt_dstypesNnames.htm

I am working on this in another branch...see ds-attribute-validation...basically over there the framework for adding better validation for any of the attributes is set up. [Here's a new discussion thread](https://github.com/zowe/vscode-extension-for-zowe/discussions/1141) where you can add anything else you think of (but I can add this suggestion you just added to that branch, you don't have to post again)

jellypuno commented:
I'm having issues with integration test with the latest commit. It was stopping here:

```
            newDSProperties.forEach((property) => {
                Object.keys(propertiesFromDsType).forEach((typeProperty) => {
                    if (typeProperty === property.key) {
                        property.value = propertiesFromDsType[typeProperty].toString();
                        property.placeHolder = propertiesFromDsType[typeProperty];
                    }
                });
            });
```

![image](https://user-images.githubusercontent.com/45042840/104748554-27b46280-5752-11eb-8ed0-422878d6ff88.png)

katelynienaber commented:
@jellypuno The integration tests should be fixed, and localization has been added for the texts in the list.

@JillieBeanSim I was thinking about adding the MF parameter names to all the items in the dropdown list, to help people search...do you think this is overkill?

Ie. all the items in the list would have the MF param name next to the text in parenthesis:
![image](https://user-images.githubusercontent.com/58596139/104852250-245bdb00-58fa-11eb-858e-eb207b94b7be.png)

jellypuno commented:
I created a Sequential Dataset with the default values and I encountered this error:

![image](https://user-images.githubusercontent.com/45042840/105891143-92965100-6010-11eb-9fc7-f6372fb0931c.png)

katelynienaber commented:
> I created a Sequential Dataset with the default values and I encountered this error:
> 
> ![image](https://user-images.githubusercontent.com/45042840/105891143-92965100-6010-11eb-9fc7-f6372fb0931c.png)
> 
> I think there's a conflict with the DIRBLK. It should be 0 instead of 5
> 
> I also notice that this happened after I created a partition dataset. Somehow the DIRBLK is transferred to my next allocation

Hey @jellypuno I am not getting this error. Could you tell me how you are going about making your DS?

I tried it by selecting Sequential Data Set in the list & allocating directly after, and also I tried going to the edit params list and allocating from there. It is getting the defaults from this file:
![image](https://user-images.githubusercontent.com/58596139/105987463-9a9cd200-609e-11eb-9e7d-b21249c42b42.png)
And there is no default DIRBLK value set for Sequential DS in that file. I can confirm no setting is being passed:
![image](https://user-images.githubusercontent.com/58596139/105987669-e2235e00-609e-11eb-8fdd-e343bc61111c.png)

JillieBeanSim commented:
instead of timing out I am now getting this error for those same integration tests
![image](https://user-images.githubusercontent.com/49491949/105997536-b9d53880-6079-11eb-823d-dbd42793b035.png)

JillieBeanSim commented:
when I try to change the spacu/alcunit to BLOCK I get these errors but also get a success message even though the data set isn't created.  I tried the value `blk`, `BLK`, `BLOCKS`, and `blks`, according to what I see in code it should be `blk` but still get the errors
![image](https://user-images.githubusercontent.com/49491949/106000528-e76fb100-607c-11eb-8a06-e3073bb0fc34.png)

katelynienaber commented:
> instead of timing out I am now getting this error for those same integration tests
> ![image](https://user-images.githubusercontent.com/49491949/105997536-b9d53880-6079-11eb-823d-dbd42793b035.png)

I think that's old, it's because I merged in master and it had a wrong variable in there. It should be fixed

JillieBeanSim commented:
I get strange login errors when refreshing the Data Set View
![image](https://user-images.githubusercontent.com/49491949/106001153-9b713c00-607d-11eb-8b4c-1fe00eb2011a.png)

I can't change the spacu/alcunit to `trk` either with this error
![image](https://user-images.githubusercontent.com/49491949/106001099-8d232000-607d-11eb-9b90-d7bbb4c54c99.png)

but when using `TRK` the data set gets created but I get this error popping up about `name not defined`.  User input should be case insensitive and not sure what is causing the error about name undefined and with the data set getting created I didn't get the info pop up about it being created
![image](https://user-images.githubusercontent.com/49491949/106001672-4124ab00-607e-11eb-8660-9431d59a6664.png)

jellypuno commented:
@katelynienaber This is how I did it:

![dsAttributes](https://user-images.githubusercontent.com/45042840/106004326-64b61880-60b3-11eb-8152-688ee827d893.gif)

katelynienaber commented:
> @katelynienaber This is how I did it:
> 
> ![dsAttributes](https://user-images.githubusercontent.com/45042840/106004326-64b61880-60b3-11eb-8152-688ee827d893.gif)

@jellypuno @JillieBeanSim Please pull again to prevent this error, it was caused by merging with master. I am checking the other issues...

katelynienaber commented:
> when I try to change the spacu/alcunit to BLOCK I get these errors but also get a success message even though the data set isn't created. I tried the value `blk`, `BLK`, `BLOCKS`, and `blks`, according to what I see in code it should be `blk` but still get the errors
> ![image](https://user-images.githubusercontent.com/49491949/106000528-e76fb100-607c-11eb-8a06-e3073bb0fc34.png)

Allocating new data set isn't a success message, it's more of an "attempting to" message. I think we should add a progress spinner here but would prefer to do so in another PR. We could change the message to something more clear...maybe something like "Attempting to allocate data set..." ?

Also in the `zowe.Create` function, [you can't pass BLOCK as an alcunit value](https://github.com/zowe/zowe-cli/blob/a9b40537e432935fda7b03e27467bb854cf3db2e/packages/zosfiles/src/methods/create/Create.ts#L178)...I don't know why XD Asking @jellypuno now...

katelynienaber commented:
> I get strange login errors when refreshing the Data Set View, these only happen if the creation fails and I refresh. If the creation was successful and I refresh view I don't see the error

I'm not getting this...I think maybe this is coming from that name error...please pull again XD

> I can't change the spacu/alcunit to `trk` either with this error

Sorry not getting this either >.<

> but when using `TRK` the data set gets created but I get this error popping up about `name not defined`. User input should be case insensitive and not sure what is causing the error about name undefined and with the data set getting created I didn't get the info pop up about it being created

This is still that error caused by merging with master, it should be fixed already.

crawr reviewed (commented):
I was testing this in Theia and I found an issue when allocating a data sets and trying to edit an attribute.

![create classic](https://user-images.githubusercontent.com/45845502/106119885-fd9d7000-6155-11eb-92ef-7cdeff8bedb2.gif)

As you can see in the gif, I received an error but the data set was allocated successfully.
<img src="https://user-images.githubusercontent.com/45845502/106121120-7bae4680-6157-11eb-871c-9b082b254053.png" width="400"/>

It is not a showstopper for me since the existing allocation still works, I just cannot edit the attributes. Perhaps the Theia specific issue can be investigated and handled in a separate PR.

katelynienaber commented:
@crawr I have opened this in an issue here: #1167

jellypuno commented:
@katelynienaber I tried to allocate a binary using default values

![allocerr](https://user-images.githubusercontent.com/45042840/106154671-220e4200-6180-11eb-8b4a-3708dbf262cd.gif)

katelynienaber commented:
@jellypuno Looks like I missed a `name` after that merge :V

Try again please. Maybe this will fix tests as well XD

jellypuno commented:
@katelynienaber Did you add a fix for the sequential dataset as well? I'm still having it even though I cloned a new copy again.

jellypuno commented:
The "allocating new dataset" message, is this just an informational message or is it a withProgress message? I'm expecting that it will be removed after the dataset was allocated.

katelynienaber commented:
> @katelynienaber Did you add a fix for the sequential dataset as well? I'm still having it even though I cloned a new copy again.

I'm not getting that error, and I thinkk I am doing the same thing as you to allocate...
![errorGif](https://user-images.githubusercontent.com/58596139/106157309-ddd07100-6182-11eb-87c1-7d7b92a9248f.gif)

katelynienaber commented:
> The "allocating new dataset" message, is this just an informational message or is it a withProgress message? I'm expecting that it will be removed after the dataset was allocated.

It's an info message, like I said to Billie, I think it should be withProgress but I would add that later and instead change the message now if it's not clear enough

jellypuno commented:
> > The "allocating new dataset" message, is this just an informational message or is it a withProgress message? I'm expecting that it will be removed after the dataset was allocated.
> 
> It's an info message, like I said to Billie, I think it should be withProgress but I would add that later and instead change the message now if it's not clear enough

I would suggest to just remove it. Because there is a Get list progress that is displayed as well. So having that allocate can be removed for now

katelynienaber commented:
@jellypuno OK I removed it!

jellypuno commented:
Error message is too vague. I executed a file creation CLI and did the same thing in Zowe Explorer. The message in the CLI is better. If we can display that instead of what we have now that would be helpful.

![image](https://user-images.githubusercontent.com/45042840/106161313-1bcf9400-6187-11eb-8c26-0f9f1f9a31df.png)

jellypuno commented:
I am not able to allocate a dataset in ZE but I can do it in CLI. I have the same attributes:

![image](https://user-images.githubusercontent.com/45042840/106162046-f1caa180-6187-11eb-8638-1a933a8ee73d.png)

jellypuno commented:
Also, Show Attributes is a CLI tag to display the attributes in the Command Line after the allocation. We don't need it and you can remove it in dialog box.

katelynienaber commented:
> Error message is too vague. I executed a file creation CLI and did the same thing in Zowe Explorer. The message in the CLI is better. If we can display that instead of what we have now that would be helpful.

CLI doesn't send us any more info. @jellypuno is talking to @t1m0thyj about this.

> I am not able to allocate a dataset in ZE but I can do it in CLI. I have the same attributes:

Fixed!

> Also, Show Attributes is a CLI tag to display the attributes in the Command Line after the allocation. We don't need it and you can remove it in dialog box.

I have removed this field as well.

jellypuno commented:
@katelynienaber thank you! don't forget to push your changes 😉

katelynienaber commented:
@jellypuno should be there now! Was just running tests

jellypuno reviewed (approved):
I think the overall functionality works for me and it looks great. 

However, I see 2 issues under this PR.
1. To get a more descriptive error message. This can be opened under CLI team.
2. To have a condition that will not display the EDIT Attributes in Theia environment.

We can open new issues for these and merge this PR in master today

JillieBeanSim commented:
Functionality looks good to me and all unit tests are passing.

> For the newly created datasets, make sure it is deleted in the posttest:integration script

not sure it this is getting deleted else where...

@katelynienaber @jellypuno I see an error here running posttest and then pretest in this PR where posttest gives this
![image](https://user-images.githubusercontent.com/49491949/106279064-d86a3980-6209-11eb-811f-d7207776e7e9.png)
which stops the script so on pretest this is the error dealing with USS since the posttest stopped and didn't continue with USS actions
![image](https://user-images.githubusercontent.com/49491949/106279246-1a937b00-620a-11eb-99ed-898f32f86f63.png)

I end up having to go in and delete USS dir manually then run pretest to have a clean testing environment for integration tests

I have this failing integration test that I haven't seen yet while testing various branches
![image](https://user-images.githubusercontent.com/49491949/106279976-482cf400-620b-11eb-862a-9fe469155c2f.png)

katelynienaber commented:
@JillieBeanSim I fixed the integration test, and yes those DS are deleted after each run, so I removed the line deleting them in the posttest.

JillieBeanSim commented:
Thanks for all of your hard work on this @katelynienaber! all expected tests and checks are passing now and the scripts are working as they should be.  Approved!
